File: video_convert_webm.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

// VideoInfo almacena información sobre un archivo de video
type VideoInfo struct {
	Width    int
	Height   int
	Duration float64
	HasAudio bool
}

// ConversionOptions almacena opciones para convertir un video
type ConversionOptions struct {
	Quality int
	Resize  string
	Crop    string
	Threads int
	Verbose bool
}

// ConversionStats almacena estadísticas de la conversión por lotes
type ConversionStats struct {
	Total int
	Exito int
	Error int
	mu    sync.Mutex
}

// Método para incrementar estadísticas de forma segura
func (stats *ConversionStats) incrementarExito() {
	stats.mu.Lock()
	defer stats.mu.Unlock()
	stats.Exito++
}

func (stats *ConversionStats) incrementarError() {
	stats.mu.Lock()
	defer stats.mu.Unlock()
	stats.Error++
}

// snakeCaseFilename convierte un nombre de archivo a snake_case
func snakeCaseFilename(filename string) string {
	base := filepath.Base(filename)
	ext := filepath.Ext(base)
	name := strings.TrimSuffix(base, ext)

	// Reemplazar caracteres no alfanuméricos con guiones bajos
	re := regexp.MustCompile(`[^a-zA-Z0-9]`)
	name = re.ReplaceAllString(name, "_")

	// Convertir camelCase a snake_case
	re = regexp.MustCompile(`([a-z0-9])([A-Z])`)
	name = re.ReplaceAllString(name, "${1}_${2}")

	// Convertir a minúsculas
	name = strings.ToLower(name)

	// Eliminar guiones bajos múltiples
	re = regexp.MustCompile(`_+`)
	name = re.ReplaceAllString(name, "_")

	// Eliminar guiones bajos al inicio o final
	name = strings.Trim(name, "_")

	return name
}

// getVideoInfo obtiene información del video usando ffprobe
func getVideoInfo(videoPath string) (*VideoInfo, error) {
	// Obtener dimensiones y duración
	cmd := exec.Command(
		"ffprobe", "-v", "error", "-select_streams", "v:0",
		"-show_entries", "stream=width,height,duration",
		"-of", "csv=p=0", videoPath,
	)

	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("error al ejecutar ffprobe: %w", err)
	}

	parts := strings.Split(strings.TrimSpace(string(output)), ",")
	if len(parts) < 3 {
		return nil, errors.New("la salida de ffprobe no contiene suficiente información")
	}

	width, err := strconv.Atoi(parts[0])
	if err != nil {
		return nil, fmt.Errorf("error al convertir ancho: %w", err)
	}

	height, err := strconv.Atoi(parts[1])
	if err != nil {
		return nil, fmt.Errorf("error al convertir alto: %w", err)
	}

	duration, err := strconv.ParseFloat(parts[2], 64)
	if err != nil {
		duration = 0
	}

	// Verificar si tiene audio
	cmdAudio := exec.Command(
		"ffprobe", "-v", "error", "-select_streams", "a",
		"-show_entries", "stream=codec_type", "-of", "csv=p=0",
		videoPath,
	)

	audioOutput, _ := cmdAudio.Output()
	hasAudio := len(strings.TrimSpace(string(audioOutput))) > 0

	return &VideoInfo{
		Width:    width,
		Height:   height,
		Duration: duration,
		HasAudio: hasAudio,
	}, nil
}

// convertToWebm convierte un video a formato WebM
func convertToWebm(inputVideo, outputPath string, opts ConversionOptions) error {
	// Verificar si el video existe
	if _, err := os.Stat(inputVideo); os.IsNotExist(err) {
		return fmt.Errorf("el archivo '%s' no existe", inputVideo)
	}

	// Obtener información del video
	videoInfo, err := getVideoInfo(inputVideo)
	if err != nil {
		return fmt.Errorf("error al obtener información del video: %w", err)
	}

	// Determinar ruta de salida
	if outputPath == "" {
		dir := filepath.Dir(inputVideo)
		filename := snakeCaseFilename(filepath.Base(inputVideo)) + ".webm"
		outputPath = filepath.Join(dir, filename)
	}

	// Preparar directorio de salida
	if err := os.MkdirAll(filepath.Dir(outputPath), 0755); err != nil {
		return fmt.Errorf("error al crear directorio de salida: %w", err)
	}

	// Convertir calidad (0-100) a bitrate aproximado (kbps)
	var bitrate int
	if opts.Quality < 30 {
		// Calidades muy bajas: 100-500 kbps
		bitrate = 100 + (400*opts.Quality)/30
	} else if opts.Quality < 70 {
		// Calidades medias: 500-2000 kbps
		bitrate = 500 + (1500*(opts.Quality-30))/40
	} else {
		// Calidades altas: 2000-6000 kbps
		bitrate = 2000 + (4000*(opts.Quality-70))/30
	}

	// Construir comando ffmpeg
	args := []string{"-y"}

	// Reducir verbosidad si no está en modo verbose
	if !opts.Verbose {
		args = append(args, "-v", "warning")
	}

	args = append(args, "-i", inputVideo)

	// Aplicar filtros si es necesario
	var filters []string

	// Filtro de recorte
	if opts.Crop != "" {
		parts := strings.Split(opts.Crop, ":")
		if len(parts) == 4 {
			x, y, w, h := parts[0], parts[1], parts[2], parts[3]
			filters = append(filters, fmt.Sprintf("crop=%s:%s:%s:%s", w, h, x, y))
		}
	}

	// Filtro de redimensionamiento
	if opts.Resize != "" {
		parts := strings.Split(opts.Resize, "x")
		if len(parts) == 2 {
			width, height := parts[0], parts[1]
			filters = append(filters, fmt.Sprintf("scale=%s:%s:force_original_aspect_ratio=decrease", width, height))
		}
	}

	// Agregar filtros al comando
	if len(filters) > 0 {
		args = append(args, "-vf", strings.Join(filters, ","))
	}

	// Configuración de codificación
	args = append(args,
		"-c:v", "libvpx-vp9",
		"-b:v", fmt.Sprintf("%dk", bitrate),
		"-deadline", "good",
		"-cpu-used", "4",
		"-pix_fmt", "yuv420p",
	)

	// Configurar número de hilos
	if opts.Threads > 0 {
		args = append(args, "-threads", strconv.Itoa(opts.Threads))
	}

	// Configuración de audio
	if videoInfo.HasAudio {
		args = append(args,
			"-c:a", "libopus",
			"-b:a", "96k",
		)
	}

	// Archivo de salida
	args = append(args, outputPath)

	// Mensaje inicial
	fmt.Printf("Convirtiendo: %s\n", filepath.Base(inputVideo))
	if opts.Verbose {
		fmt.Printf("Comando: ffmpeg %s\n", strings.Join(args, " "))
	}

	// Ejecutar comando
	cmd := exec.Command("ffmpeg", args...)
	if opts.Verbose {
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
	}

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("error durante la conversión: %w", err)
	}

	// Verificar tamaños para comparación
	inputInfo, err := os.Stat(inputVideo)
	if err != nil {
		return fmt.Errorf("error al obtener tamaño del archivo original: %w", err)
	}

	outputInfo, err := os.Stat(outputPath)
	if err != nil {
		return fmt.Errorf("error al obtener tamaño del archivo convertido: %w", err)
	}

	inputSize := float64(inputInfo.Size()) / (1024 * 1024) // MB
	outputSize := float64(outputInfo.Size()) / (1024 * 1024) // MB
	ratio := (outputSize / inputSize) * 100

	// Mostrar información de tamaño
	fmt.Printf("✓ %s - %.2f MB (%.1f%% del original)\n", filepath.Base(outputPath), outputSize, ratio)

	return nil
}

// processDirectory procesa todos los videos en un directorio
func processDirectory(inputDir, outputDir string, opts ConversionOptions, recursive bool, maxWorkers int) (*ConversionStats, error) {
	// Verificar directorio de entrada
	if _, err := os.Stat(inputDir); os.IsNotExist(err) {
		return nil, fmt.Errorf("el directorio '%s' no existe", inputDir)
	}

	// Determinar directorio de salida
	if outputDir == "" {
		outputDir = filepath.Join(inputDir, "webm")
	}

	// Crear directorio de salida si no existe
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return nil, fmt.Errorf("error al crear directorio de salida: %w", err)
	}

	// Extensiones de video soportadas
	videoExtensions := map[string]bool{
		".mp4":  true,
		".avi":  true,
		".mov":  true,
		".mkv":  true,
		".flv":  true,
		".wmv":  true,
		".webm": true,
	}

	// Encontrar todos los videos
	var videos []string

	if recursive {
		// Buscar en subdirectorios
		err := filepath.Walk(inputDir, func(path string, info fs.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if !info.IsDir() && videoExtensions[strings.ToLower(filepath.Ext(path))] {
				videos = append(videos, path)
			}
			return nil
		})
		if err != nil {
			return nil, fmt.Errorf("error al buscar videos: %w", err)
		}
	} else {
		// Buscar solo en el directorio principal
		entries, err := os.ReadDir(inputDir)
		if err != nil {
			return nil, fmt.Errorf("error al leer directorio: %w", err)
		}

		for _, entry := range entries {
			if !entry.IsDir() && videoExtensions[strings.ToLower(filepath.Ext(entry.Name()))] {
				videos = append(videos, filepath.Join(inputDir, entry.Name()))
			}
		}
	}

	if len(videos) == 0 {
		fmt.Printf("No se encontraron videos en '%s'\n", inputDir)
		return &ConversionStats{}, nil
	}

	fmt.Printf("Encontrados %d videos para procesar\n", len(videos))

	stats := &ConversionStats{
		Total: len(videos),
	}

	// Preparar canal de trabajo
	type workItem struct {
		videoPath string
		index     int
	}

	workChan := make(chan workItem, len(videos))
	for i, video := range videos {
		workChan <- workItem{video, i}
	}
	close(workChan)

	var wg sync.WaitGroup
	numWorkers := maxWorkers
	if numWorkers <= 0 {
		numWorkers = 1
	}
	if numWorkers > len(videos) {
		numWorkers = len(videos)
	}

	// Función para procesar un video
	processVideo := func(item workItem) bool {
		videoPath := item.videoPath
		relPath, err := filepath.Rel(inputDir, videoPath)
		if err != nil {
			relPath = filepath.Base(videoPath)
		}

		outputSubdir := filepath.Dir(relPath)
		fullOutputDir := filepath.Join(outputDir, outputSubdir)

		// Asegurar que existe el subdirectorio de salida
		if err := os.MkdirAll(fullOutputDir, 0755); err != nil {
			fmt.Printf("Error al crear subdirectorio: %s\n", err)
			return false
		}

		outputFile := filepath.Join(
			fullOutputDir,
			snakeCaseFilename(filepath.Base(videoPath))+".webm",
		)

		// Comprobar si el archivo ya existe y es más reciente que el original
		if info, err := os.Stat(outputFile); err == nil {
			inputInfo, err := os.Stat(videoPath)
			if err == nil {
				if info.ModTime().After(inputInfo.ModTime()) {
					fmt.Printf("Omitiendo %s - ya procesado\n", filepath.Base(videoPath))
					return true
				}
			}
		}

		// Convertir video
		err = convertToWebm(videoPath, outputFile, opts)
		if err != nil {
			fmt.Printf("Error al convertir %s: %s\n", filepath.Base(videoPath), err)
			return false
		}

		return true
	}

	// Iniciar trabajadores
	if numWorkers <= 1 {
		// Modo secuencial
		for item := range workChan {
			success := processVideo(item)
			if success {
				stats.Exito++
			} else {
				stats.Error++
			}
		}
	} else {
		// Modo paralelo
		wg.Add(numWorkers)
		for i := 0; i < numWorkers; i++ {
			go func() {
				defer wg.Done()
				for item := range workChan {
					success := processVideo(item)
					if success {
						stats.incrementarExito()
					} else {
						stats.incrementarError()
					}
				}
			}()
		}
		wg.Wait()
	}

	// Mostrar estadísticas
	fmt.Printf("\nProceso completado:\n")
	fmt.Printf("- Total procesados: %d\n", stats.Total)
	fmt.Printf("- Conversiones exitosas: %d\n", stats.Exito)
	fmt.Printf("- Errores: %d\n", stats.Error)

	return stats, nil
}

func main() {
	// Definir comandos
	fileCmd := flag.NewFlagSet("file", flag.ExitOnError)
	dirCmd := flag.NewFlagSet("dir", flag.ExitOnError)

	// Variables comunes
	var quality int
	var resize, crop string
	var verbose bool

	// Variables para comando 'file'
	fileInput := fileCmd.String("input", "", "Archivo de video de entrada")
	fileOutput := fileCmd.String("output", "", "Ruta de salida (opcional)")
	fileCmd.IntVar(&quality, "quality", 30, "Calidad del video (0-100)")
	fileCmd.StringVar(&resize, "resize", "", "Redimensionar video (formato: widthxheight)")
	fileCmd.StringVar(&crop, "crop", "", "Recortar video (formato: x:y:width:height)")
	fileCmd.BoolVar(&verbose, "verbose", false, "Mostrar información detallada")
	fileCmd.BoolVar(&verbose, "v", false, "Mostrar información detallada (forma corta)")

	// Variables para comando 'dir'
	dirInput := dirCmd.String("input", "", "Directorio de entrada")
	dirOutput := dirCmd.String("output", "", "Directorio de salida (opcional)")
	dirCmd.IntVar(&quality, "quality", 30, "Calidad del video (0-100)")
	dirCmd.StringVar(&resize, "resize", "", "Redimensionar video (formato: widthxheight)")
	dirCmd.StringVar(&crop, "crop", "", "Recortar video (formato: x:y:width:height)")
	recursive := dirCmd.Bool("recursive", false, "Buscar videos en subdirectorios")
	workers := dirCmd.Int("workers", 1, "Número máximo de trabajos en paralelo")
	dirCmd.BoolVar(&verbose, "verbose", false, "Mostrar información detallada")
	dirCmd.BoolVar(&verbose, "v", false, "Mostrar información detallada (forma corta)")

	// Verificar si hay argumentos
	if len(os.Args) < 2 {
		fmt.Println("Se requiere un subcomando: 'file' o 'dir'")
		fmt.Println("Uso:")
		fmt.Println("  webm_converter file -input <archivo> [opciones]")
		fmt.Println("  webm_converter dir -input <directorio> [opciones]")
		os.Exit(1)
	}

	// Mostrar banner
	fmt.Println("╔═══════════════════════════════════════╗")
	fmt.Println("║          WebM Converter v1.0          ║")
	fmt.Println("╚═══════════════════════════════════════╝")

	// Analizar argumentos según el subcomando
	switch os.Args[1] {
	case "file":
		fileCmd.Parse(os.Args[2:])
		if *fileInput == "" {
			fmt.Println("Error: Se requiere especificar un archivo de entrada")
			fileCmd.PrintDefaults()
			os.Exit(1)
		}

		// Validar argumentos
		if quality < 0 || quality > 100 {
			fmt.Println("Error: La calidad debe estar entre 0 y 100")
			os.Exit(1)
		}

		// Configurar opciones
		opts := ConversionOptions{
			Quality: quality,
			Resize:  resize,
			Crop:    crop,
			Verbose: verbose,
		}

		// Convertir archivo
		start := time.Now()
		err := convertToWebm(*fileInput, *fileOutput, opts)
		if err != nil {
			fmt.Printf("Error: %s\n", err)
			os.Exit(1)
		}
		elapsed := time.Since(start)
		fmt.Printf("Tiempo de conversión: %.2f segundos\n", elapsed.Seconds())

	case "dir":
		dirCmd.Parse(os.Args[2:])
		if *dirInput == "" {
			fmt.Println("Error: Se requiere especificar un directorio de entrada")
			dirCmd.PrintDefaults()
			os.Exit(1)
		}

		// Validar argumentos
		if quality < 0 || quality > 100 {
			fmt.Println("Error: La calidad debe estar entre 0 y 100")
			os.Exit(1)
		}

		// Configurar opciones
		opts := ConversionOptions{
			Quality: quality,
			Resize:  resize,
			Crop:    crop,
			Verbose: verbose,
		}

		// Procesar directorio
		start := time.Now()
		_, err := processDirectory(*dirInput, *dirOutput, opts, *recursive, *workers)
		if err != nil {
			fmt.Printf("Error: %s\n", err)
			os.Exit(1)
		}
		elapsed := time.Since(start)
		fmt.Printf("Tiempo total: %.2f segundos\n", elapsed.Seconds())

	default:
		fmt.Printf("Comando desconocido: %s\n", os.Args[1])
		fmt.Println("Use 'file' o 'dir'")
		os.Exit(1)
	}
}

File: video_convert_webm.py
#!/usr/bin/env python3
import argparse
import os
import re
import subprocess
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor


def snake_case_filename(filename):
    """Convierte un nombre de archivo a snake_case sin extensión"""
    name = Path(filename).stem
    # Reemplazar caracteres no alfanuméricos con guiones bajos
    name = re.sub(r'[^a-zA-Z0-9]', '_', name)
    # Convertir camelCase o PascalCase a snake_case
    name = re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()
    # Eliminar guiones bajos múltiples
    name = re.sub(r'_+', '_', name)
    # Eliminar guiones bajos al inicio o final
    name = name.strip('_')
    return name


def get_video_info(video_path):
    """Obtiene información del video usando ffprobe"""
    try:
        # Obtener dimensiones del video
        cmd = [
            'ffprobe', '-v', 'error', '-select_streams', 'v:0',
            '-show_entries', 'stream=width,height,duration',
            '-of', 'csv=p=0', video_path
        ]
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        width, height, duration = result.stdout.strip().split(',')

        # Verificar si tiene audio
        cmd_audio = [
            'ffprobe', '-v', 'error', '-select_streams', 'a',
            '-show_entries', 'stream=codec_type', '-of', 'csv=p=0',
            video_path
        ]
        has_audio = len(subprocess.run(cmd_audio, capture_output=True,
                                       text=True).stdout.strip()) > 0

        return {
            'width': int(width),
            'height': int(height),
            'duration': float(duration) if duration else None,
            'has_audio': has_audio
        }
    except subprocess.CalledProcessError as e:
        print(f"Error al obtener información del video: {e}")
        return None
    except ValueError as e:
        print(f"Error al procesar información del video: {e}")
        return None


def convert_to_webm(input_video, output_path=None, quality=30,
                    resize=None, crop=None, threads=None, verbose=False):
    """
    Convierte un video al formato WebM

    Args:
        input_video: Ruta del video de entrada
        output_path: Ruta del video de salida (por defecto: nombre_original.webm)
        quality: Calidad del video (0-100), donde 0 es la peor y 100 la mejor
        resize: Tuple (width, height) para redimensionar el video
        crop: Tuple (x, y, width, height) para recortar el video
        threads: Número de hilos para la codificación
        verbose: Mostrar información detallada

    Returns:
        bool: True si la conversión fue exitosa, False en caso contrario
    """
    try:
        # Verificar si el video existe
        if not os.path.exists(input_video):
            print(f"Error: El archivo '{input_video}' no existe")
            return False

        # Obtener información del video
        video_info = get_video_info(input_video)
        if not video_info:
            return False

        # Determinar ruta de salida
        if not output_path:
            directory = os.path.dirname(input_video)
            filename = f"{snake_case_filename(os.path.basename(input_video))}.webm"
            output_path = os.path.join(directory, filename)

        # Convertir calidad (0-100) a bitrate aproximado (kbps)
        # Calibración más agresiva para generar archivos más pequeños
        if quality < 30:
            # Calidades muy bajas: 100-500 kbps
            bitrate = 100 + (400 * quality / 30)
        elif quality < 70:
            # Calidades medias: 500-2000 kbps
            bitrate = 500 + (1500 * (quality - 30) / 40)
        else:
            # Calidades altas: 2000-6000 kbps
            bitrate = 2000 + (4000 * (quality - 70) / 30)

        bitrate = int(bitrate)

        # Construir comando ffmpeg con menos verbosidad
        cmd = ['ffmpeg', '-y']

        # Reducir verbosidad si no está en modo verbose
        if not verbose:
            cmd.extend(['-v', 'warning'])

        cmd.extend(['-i', input_video])

        # Aplicar filtros si es necesario
        filter_complex = []

        # Filtro de recorte
        if crop:
            x, y, w, h = crop
            filter_complex.append(f"crop={w}:{h}:{x}:{y}")

        # Filtro de redimensionamiento
        if resize:
            width, height = resize
            filter_complex.append(f"scale={width}:{height}:force_original_aspect_ratio=decrease")

        # Agregar filtros al comando
        if filter_complex:
            cmd.extend(['-vf', ','.join(filter_complex)])

        # Configuración de codificación
        cmd.extend([
            '-c:v', 'libvpx-vp9',  # Codec VP9
            '-b:v', f'{bitrate}k',  # Bitrate
            '-deadline', 'good',  # Balance entre velocidad y calidad
            '-cpu-used', '4',  # Mayor valor = más rápido, menor calidad
            '-pix_fmt', 'yuv420p'  # Formato de pixel estándar para evitar advertencias
        ])

        # Configurar número de hilos
        if threads:
            cmd.extend(['-threads', str(threads)])

        # Configuración de audio
        if video_info['has_audio']:
            cmd.extend([
                '-c:a', 'libopus',  # Codec Opus para audio
                '-b:a', '96k'  # Bitrate de audio reducido
            ])

        # Archivo de salida
        cmd.append(output_path)

        # Mensaje inicial
        print(f"Convirtiendo: {os.path.basename(input_video)}")
        if verbose:
            print(f"Comando: {' '.join(cmd)}")

        # Ejecutar comando
        process = subprocess.run(cmd,
                                 capture_output=not verbose,
                                 text=True)
        process.check_returncode()

        # Verificar tamaños para comparación
        input_size = os.path.getsize(input_video) / (1024 * 1024)  # MB
        output_size = os.path.getsize(output_path) / (1024 * 1024)  # MB
        ratio = (output_size / input_size) * 100 if input_size > 0 else 0

        # Mostrar información de tamaño
        print(f"✓ {os.path.basename(output_path)} - {output_size:.2f} MB ({ratio:.1f}% del original)")

        return True

    except subprocess.CalledProcessError as e:
        print(f"Error durante la conversión: {e}")
        if not verbose and e.stderr:
            print(f"Detalles: {e.stderr.strip()}")
        return False
    except Exception as e:
        print(f"Error inesperado: {e}")
        return False


def process_directory(input_dir, output_dir=None, quality=30, resize=None,
                      crop=None, recursive=False, max_workers=1, verbose=False):
    """
    Procesa todos los videos en un directorio

    Args:
        input_dir: Directorio de entrada
        output_dir: Directorio de salida (por defecto: input_dir/webm)
        quality: Calidad del video (0-100)
        resize: Tuple (width, height) para redimensionar
        crop: Tuple (x, y, width, height) para recortar
        recursive: Buscar videos en subdirectorios
        max_workers: Número máximo de trabajos en paralelo (default: 1)
        verbose: Mostrar información detallada

    Returns:
        dict: Estadísticas del proceso
    """
    # Verificar directorio de entrada
    if not os.path.exists(input_dir):
        print(f"Error: El directorio '{input_dir}' no existe")
        return {'total': 0, 'exito': 0, 'error': 0}

    # Determinar directorio de salida
    if not output_dir:
        output_dir = os.path.join(input_dir, 'webm')

    # Crear directorio de salida si no existe
    os.makedirs(output_dir, exist_ok=True)

    # Extensiones de video soportadas
    video_extensions = ('.mp4', '.avi', '.mov', '.mkv', '.flv', '.wmv', '.webm')

    # Encontrar todos los videos
    videos = []

    if recursive:
        # Buscar en subdirectorios
        for root, _, files in os.walk(input_dir):
            for file in files:
                if file.lower().endswith(video_extensions):
                    videos.append(os.path.join(root, file))
    else:
        # Buscar solo en el directorio principal
        videos = [os.path.join(input_dir, f) for f in os.listdir(input_dir)
                  if f.lower().endswith(video_extensions)]

    if not videos:
        print(f"No se encontraron videos en '{input_dir}'")
        return {'total': 0, 'exito': 0, 'error': 0}

    print(f"Encontrados {len(videos)} videos para procesar")

    # Función para procesar un video individual
    def process_video(video_path):
        rel_path = os.path.relpath(video_path, input_dir)
        output_subdir = os.path.dirname(rel_path)

        # Crear subdirectorio en la salida si es necesario
        full_output_dir = os.path.join(output_dir, output_subdir)
        os.makedirs(full_output_dir, exist_ok=True)

        output_file = os.path.join(
            full_output_dir,
            f"{snake_case_filename(os.path.basename(video_path))}.webm"
        )

        # Comprobar si el archivo ya existe y es más reciente que el original
        if os.path.exists(output_file):
            input_time = os.path.getmtime(video_path)
            output_time = os.path.getmtime(output_file)
            if output_time >= input_time:
                print(f"Omitiendo {os.path.basename(video_path)} - ya procesado")
                return True

        # Convertir video
        success = convert_to_webm(
            video_path,
            output_file,
            quality,
            resize,
            crop,
            threads=1,  # Un solo hilo por worker
            verbose=verbose
        )

        return success

    # Procesar videos secuencialmente o en paralelo
    results = []
    if max_workers <= 1:
        # Procesamiento secuencial
        for video in videos:
            results.append(process_video(video))
    else:
        # Procesamiento en paralelo (limitado)
        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            futures = [executor.submit(process_video, video) for video in videos]
            for future in futures:
                results.append(future.result())

    # Estadísticas
    stats = {
        'total': len(videos),
        'exito': sum(results),
        'error': len(videos) - sum(results)
    }

    print(f"\nProceso completado:")
    print(f"- Total procesados: {stats['total']}")
    print(f"- Conversiones exitosas: {stats['exito']}")
    print(f"- Errores: {stats['error']}")

    return stats


def main():
    parser = argparse.ArgumentParser(description='Convertir videos a formato WebM')
    subparsers = parser.add_subparsers(dest='mode', help='Modo de operación')

    # Subparser para modo archivo único
    file_parser = subparsers.add_parser('file', help='Convertir un archivo')
    file_parser.add_argument('input', help='Archivo de video de entrada')
    file_parser.add_argument('--output', help='Ruta de salida (opcional)')

    # Subparser para modo directorio
    dir_parser = subparsers.add_parser('dir', help='Convertir todos los videos en un directorio')
    dir_parser.add_argument('input_dir', help='Directorio de entrada')
    dir_parser.add_argument('--output-dir', help='Directorio de salida (opcional)')
    dir_parser.add_argument('--recursive', action='store_true',
                            help='Buscar videos en subdirectorios')
    dir_parser.add_argument('--workers', type=int, default=1,
                            help='Número máximo de trabajos en paralelo (default: 1)')

    # Argumentos comunes
    for p in [file_parser, dir_parser]:
        p.add_argument('--quality', type=int, default=30,
                       help='Calidad del video (0-100, default: 30)')
        p.add_argument('--resize', type=str,
                       help='Redimensionar video (formato: widthxheight)')
        p.add_argument('--crop', type=str,
                       help='Recortar video (formato: x:y:width:height)')
        p.add_argument('--verbose', '-v', action='store_true',
                       help='Mostrar información detallada')

    args = parser.parse_args()

    # Validar argumentos
    if args.mode not in ['file', 'dir']:
        parser.print_help()
        return

    # Mostrar banner
    print("╔═══════════════════════════════════════╗")
    print("║          WebM Converter v1.0          ║")
    print("╚═══════════════════════════════════════╝")

    # Procesar argumentos de redimensionado
    resize = None
    if hasattr(args, 'resize') and args.resize:
        try:
            width, height = map(int, args.resize.split('x'))
            resize = (width, height)
        except ValueError:
            print("Error: Formato de resize incorrecto. Usar widthxheight (ej: 640x480)")
            return

    # Procesar argumentos de recorte
    crop = None
    if hasattr(args, 'crop') and args.crop:
        try:
            x, y, width, height = map(int, args.crop.split(':'))
            crop = (x, y, width, height)
        except ValueError:
            print("Error: Formato de crop incorrecto. Usar x:y:width:height (ej: 10:10:640:480)")
            return

    # Validar calidad
    if args.quality < 0 or args.quality > 100:
        print("Error: La calidad debe estar entre 0 y 100")
        return

    # Ejecutar en modo archivo único
    if args.mode == 'file':
        if not os.path.exists(args.input):
            print(f"Error: El archivo '{args.input}' no existe")
            return

        convert_to_webm(
            args.input,
            args.output,
            args.quality,
            resize,
            crop,
            verbose=args.verbose
        )

    # Ejecutar en modo directorio
    elif args.mode == 'dir':
        if not os.path.exists(args.input_dir):
            print(f"Error: El directorio '{args.input_dir}' no existe")
            return

        process_directory(
            args.input_dir,
            args.output_dir,
            args.quality,
            resize,
            crop,
            args.recursive,
            args.workers,
            args.verbose
        )


if __name__ == "__main__":
    main()